core/Pos/Binary/Core/Ssc.hs

```diff
@@ -80,12 +80,7 @@ encodeVssCertificates = encode . HS.fromList . toList
 decodeVssCertificates :: HasConfiguration => Decoder s VssCertificatesMap
 decodeVssCertificates = do
     certs <- toList <$> decode @(HashSet VssCertificate)
-    -- If the attacker creates two certs that are different but have the
-    -- same 'vcSigningKey', it's bad because then we lose canonicity (only
-    -- one cert will be present in resulting map and the attacker can set
-    -- the other cert to be anything at all). 'mkVssCertificatesMap' checks
-    -- that all certificates have distinct keys, so we can safely use it.
-    toCborError $ mkVssCertificatesMap certs
+    pure $ mkVssCertificatesMap certs
```

Comment:
We require canonicity from all `Bi` instances (more details can be found [here](http://cardano-dev.askbot.com/question/3/what-is-canonicity-in-the-context-of-bi-and-what-is-it-for/)).

Now `mkVssCertificatesMap` doesn't check anything. It means that if I encode a list with two certificates which have same `vsSigningKey`s, they will be decoded to the same value. Please fix this instance or provide a good reason not to do it (I don't see one).

Reply:
Oh is `mkVssCertificatesMap` responsible for enforcing canonicity? Shouldn't the `Bi (HashSet VssCertificate)` instance do that?

Reply:
Indeed it does, by way of `decodeSetSkel`.

Reply:
Instance `Bi (HashSet VssCertificate)` ensures canonicity of `HashSet VssCertificate` decoding.
Suppose there are two different `VssCertificate`s with same `vcSigningKey`: `cert1` and `cert2`.
```
let bs1 = encode @(HashSet VssCertificate) [cert1, cert2]
    bs2 = encode @(HashSet VssCertificate) [cert1]
    Right m1 = decode @VssCertificatesMap bs1
    Right m2 = decode @VssCertificatesMap bs2
```

Now `bs1 ≠ bs2`, but `m1 = m2`.

Reply:
`encode @(HashSet VssCertificate) [cert1, cert2]` is ill-typed, no? Shouldn't it be `encode @(HashSet VssCertificate) (fromList [cert1, cert2])`? Since `fromList [cert1, cert2] == fromList [cert1]`, `bs1 = bs2`. What am I missing?

Reply:
Yes, I forgot `fromList`, sorry.
> Since fromList [cert1, cert2] == fromList [cert1]

No, why? As I said, certificates are different. Only `vcSigningKey`s are same.

Reply:
> No, why? As I said, certificates are different. Only vcSigningKeys are same.

I thought maybe the `Eq` instance had certs with equal signing keys be equal. But I see now: you encode a `HashSet`, but decode it as a `VssCertificatesMap`. Is that OK to do?

```
let bs1 = encode @(VssCertificatesMap) (fromList [cert1, cert2])
    bs2 = encode @(VssCertificatesMap) (fromList [cert1])
    Right m1 = decode @VssCertificatesMap bs1
    Right m2 = decode @VssCertificatesMap bs2
```

Wouldn't that satisfy canonicity? `b1 = b2` right?

Reply:
Yes indeed, I believe canonicity is maintained for `VssCertificatesMap`, but sure, if you encode a set that has duplicate keys, and then try to decode it to a *different type*, you may not get what you expect.

Reply:
> Is that OK to do?

I think so. Keys in `VssCertificatesMap` can be calculated from values, so why not?
Anyway, even if it's not ok, that's how it's done now and we should preserve it for compatibility with mainnet.

In order to satisfy canonicity `decodeVssCertificatesMap` should check that all `vsSigningKey`s are different, as it was doing before.

Reply:
> I think so.

What's the rule for canonicity?

```
decode bs1 = decode bs2 ==> bs1 = bs2
```

Implicit here is the type of `decode`. To be more concrete:

```
(decode bs1 :: a) = (decode bs2 :: a) ==> bs1 = bs2
```

I don't believe canonicity requires that

```
(mkVssCerticiatesMap (decode bs1 :: HashSet VssCerticiate)) = (decode bs2 :: VssCertificatesMap) ==> bs1 = bs2
```

Reply:
Hm, if it doesn't require the last property, then nevermind. I've asked @neongreen to comment, since he wrote this comment initially.

Reply:
@neongreen answered to me that he initially wrote incorrect definition of canonicity and proper definition is more strict. Specifically, two bytestrings `a` and `b` should deserialize to the same value of given type if and only if the are equal.
So this function violates canonicity.